SOLR-18286: Add Changelog Types Reference to developer documentation (#4530)
apache/solr: dev-docs/changelog.adoc

    url: https://issues.apache.org/jira/browse/SOLR-3333
----

=== Changelog Types Reference

[cols="1h,2,3", options="header"]
|===
| Type | Description | When to Use (Examples)

| added

> | For completely new features, capabilities, or opt-in functionality.

Reviewer: ```suggestion
| For changes requiring a user to take action to use (opt-in).  Could be completely new features or simply new configuration values on existing features.  Documented.
```

Emphasizing opt-in as the first order gating category.

| Adding a new API endpoint, a new query shortcut, or a new optional configuration parameter.

| changed
| For changes to existing functionality that alter current behavior.

> | Modifying an existing default behavior, updating an internal dependency, or refactoring a core algorithm.

Reviewer: ```suggestion
| Modifying behavior or performance of existing requests/configuration.
```

No: Not usually a dependency update.  Not usually a refactoring.   Both of those are usually "other".

Reviewer 2: So @abumarjikar should add a "other" category?   I wonder if dependency updates are common enough that they merit their own category?

Author: added other and corrected the table based on the types mentioned in gradle file

| deprecated
| For features or APIs that are slated for removal in future releases.
| Marking an old configuration option or an old method as deprecated.

| removed
| For features or APIs that have been completely removed from the codebase.
| Deleting a feature that was previously marked as deprecated.

| fixed
| For any bug fixes or corrections to unintended behavior.
| Fixing a `NullPointerException`, correcting data corruption issues, or resolving UI glitches.

> | security

Reviewer: @janhoy lets drop this.  This is an independent cross-cutting concern.  Perhaps we might have a separate loose tagging mechanism with values like "security", "dependency", "perf" -- or just not bother.

    url: https://issues.apache.org/jira/browse/SOLR-3333
----

=== Changelog Types Reference

[cols="1,2,3",options="header"]
|===
| Type | Description | When to Use (Examples)

| `added`
| For changes requiring a user to take action to use (opt-in).
| Could be completely new features or simply new configuration values on existing features. Typically documented in the Ref Guide.

| `changed`
| For improvements; not opt-in.
| Modifying behavior or performance of existing requests/configuration.

| `fixed`
| For improvements that are deemed to have fixed buggy behavior.
| Fixing a `NullPointerException`, correcting data corruption issues, or resolving UI glitches.

| `deprecated`
| For marking things deprecated.
| Declaring a configuration option, class, or API endpoint as deprecated ahead of a future removal.

| `removed`
| For code removed.
| Removing a previously deprecated parameter, class, or legacy module.

| `dependency_update`
| For updates to dependencies.
| Bumping third-party library versions (e.g., Lucene, Jetty, Jackson).

| `other`

> | For anything else, like large/significant refactorings, build changes, test infrastructure, or documentation. Most such changes are too small/minor to bother with a changelog entry.

Reviewer: ```suggestion
| For anything else. Most such changes are too small/minor to bother with a changelog entry.
```

Author: Will reflect in next commit

=== Changelog Types Reference

[cols="1,2,3",options="header"]
|===
| Type | Description | When to Use (Examples)

| `added`
| For changes requiring a user to take action to use (opt-in).
| Could be completely new features or simply new configuration values on existing features. Typically documented in the Ref Guide.

| `changed`
| For improvements; not opt-in.
| Modifying behavior or performance of existing requests/configuration.

| `fixed`
| For improvements that are deemed to have fixed buggy behavior.
| Fixing a `NullPointerException`, correcting data corruption issues, or resolving UI glitches.

| `deprecated`
| For marking things deprecated.
| Declaring a configuration option, class, or API endpoint as deprecated ahead of a future removal.

| `removed`
| For code removed.
| Removing a previously deprecated parameter, class, or legacy module.

| `dependency_update`
| For updates to dependencies.
| Bumping third-party library versions (e.g., Lucene, Jetty, Jackson).

| `other`
| For anything else, like large/significant refactorings, build changes, test infrastructure, or documentation.

> | Most such changes are too small/minor to bother with a changelog entry.

Reviewer: this section has 2 columns -- description then examples.  The information on this row for "other" isn't organized this way but could be easily.  Basically "for anything else" is the description (adding the "Most such changes are too small/minor to bother with a changelog entry." to that part as well if it can fit, and then the examples obviously in the next column.

Author: Done, will reflect the changes in next commit.
